numpy/_core/src/umath/loops_arithmetic.dispatch.cpp

```diff
@@ -0,0 +1,281 @@
+#define _UMATHMODULE
+#define _MULTIARRAYMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include "hwy/highway.h"
+#include "loops_utils.h"
+#include "loops.h"
+#include <cstring> // for memcpy
+#include "fast_loop_macros.h"
+#include <limits>
+#include "simd/simd.h"
+#include "lowlevel_strided_loops.h"
+#include "numpy/npy_math.h"
+#include <cstdio>
+
+using namespace hwy::HWY_NAMESPACE;
+
+#ifndef NPY_CPU_DISPATCH_CURFX
+    #define NPY_CPU_DISPATCH_CURFX(FN) FN##_AVX2, FN##_SSE41, FN##_SSE2, FN##_NEON, FN##_BASELINE
+#endif
+
+
+HWY_BEFORE_NAMESPACE();
+namespace HWY_NAMESPACE {
+
+
+// Helper function to set float status
+inline void set_float_status(bool overflow, bool divbyzero) {
+    if (overflow) {
+        npy_set_floatstatus_overflow();
+    }
+    if (divbyzero) {
+        npy_set_floatstatus_divbyzero();
+    }
+}
+
+// Signed integer division
+template <typename T>
+void simd_divide_by_scalar_contig_signed(T* src, T scalar, T* dst, npy_intp len) {
+    using D = ScalableTag<T>;
+    const D d;
+    const size_t N = Lanes(d);
+
+    bool raise_overflow = false;
+    bool raise_divbyzero = false;
+
+    if (scalar == 0) {
+        for (npy_intp i = 0; i < len; i += N) {
+            const size_t num = HWY_MIN(static_cast<size_t>(len - i), N);
+            StoreN(Zero(d), d, dst + i, num);
+        }
+        raise_divbyzero = true;
+    } else if (scalar == 1) {
+        // Special case for division by 1
+        memcpy(dst, src, len * sizeof(T));
+    } else if (scalar == static_cast<T>(-1)) {
+        const auto vec_min_val = Set(d, std::numeric_limits<T>::min());
+        for (npy_intp i = 0; i < len; i += N) {
```

Comment:
I strongly recommend splitting up loops into the main body, and a tail. The LoadN/StoreN functions are quite expensive, so in the main loop they should be replaced with LoadU/StoreU.

Reply:
Ack Done.